### PR TITLE
jrnl 2.2.0

### PR DESCRIPTION
**Pull request by wren**
This is an update to PR #50304 to add more comprehensive tests.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/jrnl.rb
Comment (SMillerDev) [inline Formula/jrnl.rb]: Please do all of this just in ruby below. The current approach makes it difficult to debug where it failed. Preferably there would be as few scripts as possible.

Comment (wren) [reply Formula/jrnl.rb]: Hi, I've updated the PR with your feedback. I've opted to keep using expect since our project has much more knowledge of cli tools than Ruby, so this will be easier to maintain.

File: Formula/jrnl.rb
Comment (zbeekman) [inline Formula/jrnl.rb]: @wren While we appreciate your perspective and greatly value your desire to help maintain the formula, the goal is not to *need* you or other upstream folks to maintain this. That’s why it would be far preferable to not write out long complicated shell scripts for the test. You can still use `expect`, and can ack/ag/rg on `expect` in the Formula directory to find other, better examples of how to do this using Homebrew’s DSL.

**Review by zbeekman — CHANGES_REQUESTED** [review]

File: Formula/jrnl.rb
Comment (SMillerDev) [reply Formula/jrnl.rb]: It's a homebrew formula though, you're not a maintainer of homebrew. I'd prefer it if it were easier for homebrew maintainers and contributors to work with.

File: Formula/jrnl.rb
Comment (wren) [reply Formula/jrnl.rb]: @zbeekman Hi! I'm really sorry for the back and forth on this. I haven't worked with a Homebrew formula before, so I really appreciate your and @SMillerDev's patience.

I looked at the other formulas that use `expect`, and I'm not really clear on what makes the script in this PR complicated and unmaintainable, but makes the other scripts already in homebrew core acceptable.

Is it the explicit timeouts on each expect statement? I did this so that the tests will fail quickly (in a few seconds), as opposed to waiting for your CI's full timeout interval. I'm happy to take it out, if that's the problem. But I'm just not clear what I need to change from looking at the other formulas (they all just spawn a few commands and send a few keystrokes as far as I can see).

Comment (zbeekman) [reply Formula/jrnl.rb]: Perhaps @SMillerDev disagrees with me, but my problem with the current tests are:
1. There should not be any bash scripts written out
2. If using `expect` it should not be a shell script and you should use the `#!/usr/bin/expect -f` shebang and write tests directly in expect.
3. The new test is too complicated
4. I thought the [old tests were fine](https://github.com/Homebrew/homebrew-core/pull/50593/commits/241616e2504ce4bfcf4408fa28f1e8510beda897#diff-f234881e1eaa994e2db5d570060372a7L78-L87) and don't understand why they were initially removed in 241616e2504ce4bfcf4408fa28f1e8510beda897.

Comment (wren) [reply Formula/jrnl.rb]: Hi @zbeekman,

I can see what you're saying. Maybe I can go through each of my decisions on the things you mention above, and you can tell me if it's worth keeping, or if you still think I should remove it.

1. **Bash script written out** - In v2, jrnl changed where it stores its config file to the `$XDG_CONFIG_HOME` directory. I added a shell script to be explicit about where it should store the config file, so that we can confirm it ended up in the correct location. The intent is to ensure that the file is written to the testing directory, and also to be know where to look for it when confirming it's presence in the later test. Maybe this isn't needed. I don't know if the `$XDG_CONFIG_HOME` environment variable is set in the test environment, or maybe there's a Ruby way to set an env var for the expect script. Or maybe it'll just end up in the correct directory? I just want to be clear that the only thing the shell script does is define the env vars, and run the expect script. It's not doing anything else.

2. **Create tests directly in expect** - I'm not clear what you're saying on this one. As far as I can tell, I'm creating an expect file, with the correct shebang, and running it with `expect filename`. I'm not sure what you want me to change on this. Is it because it's being invoked by the shell script (as mentioned above)?

3. **New test is too complicated** - Jrnl has three core pieces of functionality. Writing a journal, reading a journal, and encrypting a journal. The problem with the old tests is that they only tested two of those three core pieces of functionality, and it did so in a single command. This was not only incomplete, but made it difficult to tell what had failed. The new tests test each piece of functionality, with one command each. The expect script issues exactly three commands, one for each piece of functionality (read, write, encrypt). Maybe that's overkill, but if these three commands succeed, anyone updating the formula can be very confident that jrnl will still actually work after the change.

4. **Old tests** - I agree that the old tests seemed fine on the surface. But we learned the hard way a few weeks ago that they're not enough. They failed to prevent a [breaking change](https://github.com/Homebrew/homebrew-core/issues/47274) from happening to the jrnl formula. Someone edited the formula, saw that the tests passed, and so the change was merged in (which is the expected flow). This is exactly what I'm trying to prevent from happening again, and that's why I've expanded the tests. The entire set of tests still only takes a second or so to run, so I didn't think the additional tests would be a big deal.

5. **Explicit timeouts in expect** - You didn't mention this one, but I'm not sure if this is what you're referring to when you say that the tests are too complicated. There are two ways to write expect scripts:

```sh
timeout -1 #forever timeout
spawn foo
expect "bar"
send -- "baz"
```

and

```sh
timeout 3 #small timeout
spawn foo
expect {
  "bar" { send -- "baz" }
  timeout { exit 1  } # or whatver you want instead of exit
}
```

The main difference is that that first one (with `timeout -1`) is maybe easier to read. The downside is that it will never actually time out. Your CI's timeout (which seems to be about 5 minutes) will always occur first, and any and all failures will always take the full time to fail (even if the failure happens on the first line).

The second one (with `timeout 3`), has a bit more syntax around it, since we're being explicit about each timeout instead of relying on a global one. This version is maybe a bit harder to read. The upside, though, is that it will always either pass or fail in a few seconds at most. No CI timeout required.

That's the only real difference between the scripts, though. They're both essentially just interacting with the cli through expect.

And to be clear, the jrnl tests in this PR are still only issuing three commands and sending keystrokes for the prompts. There is no additional or more complicated functionality. It's still only using `spawn`, `send`, and `expect`, which are the same basic commands the original tests, and the other formulas use.

So, I'm glad to change back to the `timeout -1` format if that's preferred, but I just wanted to be clear that it's there to save time on the CI, and for anyone working with the formula or waiting on the tests.

File: Formula/jrnl.rb
Comment (SMillerDev) [inline Formula/jrnl.rb]: This should just be ruby

**Comment by micahellison** [comment]
Looks like the Jenkins server failed because it was having a hard time connecting to Github. Is there a way to re-run this latest build?

**Comment by SMillerDev** [comment]
We're having issues with the CI network at the moment, someone will retrigger when it's resolved.

**Comment by wren** [comment]
Assuming the tests pass, is there any other feedback from a code review standpoint?

**Comment by Bo98** [comment]
Thanks @wren.

**Comment by ergl** [comment]
Upgrading `jrnl` from version `1.9.8_2` to this one `2.2.0` is causing some errors:

```
==> Downloading https://github.com/jrnl-org/jrnl/releases/download/v2.2/jrnl-2.2
Already downloaded: /Users/<snip>/Library/Caches/Homebrew/downloads/4535f967695ce9b13154c5ece7ed64008a67b6b1b2093f52a241b511ac0177b4--jrnl-2.2-brew.tar.gz
Error: An exception occurred within a child process:
  NoMethodError: undefined method `delete' for nil:NilClass
```

I don't know if this is related to jrnl or brew. Happy to open an issue where it corresponds

**Comment by Bo98** [comment]
Can I see the full stack trace?

**Comment by Bo98** [comment]
See https://github.com/Homebrew/brew/issues/7115

**Comment by ergl** [comment]
Thanks for the catch, it was related to that. Waiting for https://github.com/Homebrew/brew/pull/7119 to check that everything works.

**Pull request closed without merging**
